### PR TITLE
feat!: Introduce Definitions folder

### PR DESCRIPTION
**Pull request by asalzburger**
This PR introduces a new `Definitions` folder, separating the `Utilities` from the actual definitions.
It also splits the algebra definitions from enum definitions and other parameters.

It is the first step to prepare building Acts (or well defined blocks of Acts) with float/double precision.

Consequently it removes the `Vector3F` etc. definitions, they were practically not used within the project.

BREAKING CHANGE:
Introduction of a new `Definitions` folder and hence change of include structure, which contains:
- `Algebra.hpp` - defining the EDM/Geometry basic algebra objects and the basic Scalar type
- `Common.hpp` - some common definitions
- `Alignment.hpp` - the alignment definitions
- `Units.hpp` - the unit definitions
- `TrackParameterization.hpp` - the free/bound parameter definitions

With this change, the `Core` library builds with `float` precision, defined by a new `ACTS_CUSTOM_SCALARTYPE` variable.

### CODE REVIEW

**Comment by asalzburger** [comment]
@paulgessinger could you have a quick look how I did the Cmake changes?

I tried and it compiles fine (with float) if I do:

```console
 cmake -S . -B ../builds/build-mathlib-float -DACTS_CUSTOM_SCALARTYPE=float
```

If I add the UNIT_TESTS I get a hell lot of compiler errors (expected), which I will address in a second PR.

**Comment by paulgessinger** [comment]
Hm, so this is a combination of the floating point configuration change, and  the renaming / restructuring of the definitions. Is this intrinsically linked, or do you reckon it could be split up further?

I only see two CMake changes. Do you not register `ACTS_CUSTOM_SCALARTYPE` as an option? And what / where is the default?

File: Core/include/Acts/Definitions/Algebra.hpp
Comment (paulgessinger) [inline Core/include/Acts/Definitions/Algebra.hpp:31]: Ah I see the default is here. It might be 'better' if we just put the default in a CMake option, and just do 

```cpp
using ActsScalar = ACTS_CUSTOM_SCALARTYPE;
```

unconditionally. Then the definition and default are all in one place, and it should effectively do the same thing.

**Comment by paulgessinger** [comment]
I think #600 **needs** to be ready to go so we can merge this one and #600 back-to-back. I really don't want the CI to be all-red, or disable all the unit tests altogether. I'm ok with them going in right after one another if we're sure that applies cleanly and keeps master green.

File: Core/include/Acts/Definitions/Algebra.hpp
Comment (asalzburger) [reply Core/include/Acts/Definitions/Algebra.hpp:31]: I see, I will change that accordingly.

**Comment by asalzburger** [comment]
> I think #600 **needs** to be ready to go so we can merge this one and #600 back-to-back. I really don't want the CI to be all-red, or disable all the unit tests altogether. I'm ok with them going in right after one another if we're sure that applies cleanly and keeps master green.

Absolutely right, I will work on #600 - the plugin CI was not yet updated.

File: Core/include/Acts/Definitions/Algebra.hpp
Comment (msmk0) [reply Core/include/Acts/Definitions/Algebra.hpp:31]: To summarize my argument for leaving the current implementation as-is: this would introduce (for the first time) a non-optional preprocessor definition without which the code can not be used. This adds a (small) complication without giving us any functionality or feature. The behavior of the code would be identical regardless of the way it is implemented, so I see no necessity to change it.

While CMake can set the required default value for us, I can image that someone installs Acts as an external dependency via CMake, but then use a different build system for their own software. By switching to a required preprocessor definition, this add additional road-blocks. I agree that these are only small obstacles, but it feels to me more fragile than the current implementation.

I wanted to mention my arguments here, but I will leave it up to @asalzburger to decide.

**Review by msmk0 — CHANGES_REQUESTED** [review]
@asalzburger I am happy with the overall idea and implementation. I added a few minor consistency fixes here and there.

You should consider to rename `Definitions/Parameters.hpp` to `Definitions/TrackParameters.hpp` for consistency with our naming scheme for the track edm classes. This has the problem that it collides with the `EventData/TrackParameters.hpp` file that defines charged parameter types. The later file might possibly be renamed to `EventData/SinglyChargedTrackParameters.hpp` for consistency with `EventData/NeutralTrackParameters.hpp`.

I like the removal of the unnecessary algebra typedefs. For future PRs I would have preferred to have them in a separate changeset, though.

**Comment by asalzburger** [comment]
> @asalzburger I am happy with the overall idea and implementation. I added a few minor consistency fixes here and there.
> 
> You should consider to rename `Definitions/Parameters.hpp` to `Definitions/TrackParameters.hpp` for consistency with our naming scheme for the track edm classes. This has the problem that it collides with the `EventData/TrackParameters.hpp` file that defines charged parameter types. The later file might possibly be renamed to `EventData/SinglyChargedTrackParameters.hpp` for consistency with `EventData/NeutralTrackParameters.hpp`.
> 
> I like the removal of the unnecessary algebra typedefs. For future PRs I would have preferred to have them in a separate changeset, though.

What if we call this:
`Definitions/Parametrisation.hpp`  

That's what it is, right?

File: Core/include/Acts/Propagator/ConstrainedStep.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/ConstrainedStep.hpp:23]: Fine with that change, will apply it.

File: Core/include/Acts/Surfaces/AnnulusBounds.hpp
Comment (asalzburger) [reply Core/include/Acts/Surfaces/AnnulusBounds.hpp:201]: I actually plan to do two cleanup PRs after this and #600  

- reduce the use of explicit `Eigen::` in our repository in favour of Acts (there's the nuclear option PR #595 for discussion open)
- general cleanup of `<>` includes versus `""` as in your comment below.

Would you mind if we did that in a separate one?

File: docs/core/definitions/algebra.rst
Comment (asalzburger) [reply docs/core/definitions/algebra.rst:4]: Fixed.

File: Fatras/include/ActsFatras/EventData/Hit.hpp
Comment (asalzburger) [reply Fatras/include/ActsFatras/EventData/Hit.hpp:29]: It would need potentially `cast` operations, so I would keep it like this for the moment.

**Comment by msmk0** [comment]
> What if we call this:
> `Definitions/Parametrisation.hpp`

`Definitions/TrackParametrisation.hpp`, maybe?

**Comment by asalzburger** [comment]
> > What if we call this:
> > `Definitions/Parametrisation.hpp`
> 
> `Definitions/TrackParametrisation.hpp`, maybe?

So it be.

File: Fatras/include/ActsFatras/EventData/Hit.hpp
Comment (msmk0) [reply Fatras/include/ActsFatras/EventData/Hit.hpp:29]: Ok, understood.

File: Core/include/Acts/Surfaces/AnnulusBounds.hpp
Comment (msmk0) [reply Core/include/Acts/Surfaces/AnnulusBounds.hpp:201]: We can do the changes that you mention in separate PRs, but the suggested changes here and further down do not fit in either category. I just suggest to use the appropriate type for the weight.

**Comment by msmk0** [comment]
A further thought: should we also use `Definitions/AlignmentParametrization.hpp`?

**Comment by asalzburger** [comment]
> A further thought: should we also use `Definitions/AlignmentParametrization.hpp`?

Not so sure, as it might have also other Alignment definitions in the future, I would leave it as is.

File: Fatras/include/ActsFatras/EventData/Particle.hpp
Comment (asalzburger) [reply Fatras/include/ActsFatras/EventData/Particle.hpp:26]: Same comment as above.

File: Core/src/Surfaces/AnnulusBounds.cpp
Comment (asalzburger) [reply Core/src/Surfaces/AnnulusBounds.cpp:355]: ok.

File: Core/include/Acts/Surfaces/AnnulusBounds.hpp
Comment (asalzburger) [reply Core/include/Acts/Surfaces/AnnulusBounds.hpp:204]: ok.

File: Examples/Detectors/DD4hepDetector/include/ActsExamples/DD4hepDetector/DD4hepDetectorOptions.hpp
Comment (asalzburger) [reply Examples/Detectors/DD4hepDetector/include/ActsExamples/DD4hepDetector/DD4hepDetectorOptions.hpp:13]: See above.

File: Examples/Io/Csv/src/CsvParticleWriter.cpp
Comment (asalzburger) [reply Examples/Io/Csv/src/CsvParticleWriter.cpp:13]: See above.

File: Examples/Io/Csv/src/CsvParticleReader.cpp
Comment (asalzburger) [reply Examples/Io/Csv/src/CsvParticleReader.cpp:14]: See above.

File: Examples/Run/Generators/ParticleGun.cpp
Comment (asalzburger) [reply Examples/Run/Generators/ParticleGun.cpp:17]: See above.

**Comment by msmk0** [comment]
@asalzburger Wikipedia says the common spelling would be [`Parametrization`](https://en.wikipedia.org/wiki/Parametrization) no `Parameterization`

**Review by msmk0 — APPROVED** [review]
@asalzburger Looks good. I'll merge it in once the CI runs through.

**Pull request merged**
